File: proj/CoreLibrary/src/com/jasonzqshen/familyaccounting/core/utils/CurrencyAmount.java
package com.jasonzqshen.familyaccounting.core.utils;

import java.io.Serializable;

import com.jasonzqshen.familyaccounting.core.exception.format.CurrencyAmountFormatException;

/**
 * currency amount, ~.2
 * 
 * @author I072485
 * 
 */
public class CurrencyAmount implements Comparable<CurrencyAmount>, Serializable {

	/**
	 * 
	 */
	private static final long serialVersionUID = -4225306075522416102L;
	private int _value;

	public CurrencyAmount() {
		_value = 0;
	}

	public CurrencyAmount(Double value) {
		set(value);
	}

	public CurrencyAmount(CurrencyAmount amount) {
		if (amount == null) {
			_value = 0;
			return;
		}

		_value = amount._value;
	}

	/**
	 * +=
	 * 
	 * @param amount
	 */
	public void addTo(CurrencyAmount amount) {
		_value += amount._value;
	}

	/**
	 * -=
	 * 
	 * @param amount
	 */
	public void minusTo(CurrencyAmount amount) {
		_value -= amount._value;
	}

	/**
	 * 
	 * @param amount
	 */
	public void set(CurrencyAmount amount) {
		if (amount == null) {
			return;
		}

		_value = amount._value;
	}

	public void set(double value) {
		_value = (int) (value * 100);
	}

	@Override
	public int compareTo(CurrencyAmount arg0) {
		return this._value - arg0._value;
	}

	/**
	 * == 0
	 * 
	 * @return
	 */
	public boolean isZero() {
		return this._value == 0;
	}

	/**
	 * < 0
	 * 
	 * @return
	 */
	public boolean isNegative() {
		return this._value < 0;
	}

	/**
	 * 
	 */
	public void negate() {
		this._value = 0 - this._value;
	}

	/**
	 * oper1 + oper2
	 * 
	 * @param oper1
	 * @param oper2
	 * @return
	 */
	public static CurrencyAmount add(CurrencyAmount oper1, CurrencyAmount oper2) {
		CurrencyAmount cur = new CurrencyAmount();
		cur.addTo(oper1);
		cur.addTo(oper2);

		return cur;
	}

	/**
	 * oper1 - oper2
	 * 
	 * @param oper1
	 * @param oper2
	 * @return
	 */
	public static CurrencyAmount minus(CurrencyAmount oper1,
			CurrencyAmount oper2) {
		CurrencyAmount cur = new CurrencyAmount();
		cur.addTo(oper1);
		cur.minusTo(oper2);

		return cur;
	}

	@Override
	public boolean equals(Object obj) {
		if (!(obj instanceof CurrencyAmount)) {
			return false;
		}
		CurrencyAmount amount = (CurrencyAmount) obj;
		return this._value == amount._value;
	}

	@Override
	public String toString() {
		double value = this.toNumber();
		return String.format("%.2f", value);
	}

	/**
	 * to number
	 * 
	 * @return
	 */
	public double toNumber() {
		return _value / 100.0;
	}

	/**
	 * parse value
	 * 
	 * @param value
	 * @return
	 * @throws CurrencyAmountFormatException
	 */
	public static CurrencyAmount parse(String value)
			throws CurrencyAmountFormatException {
		try {
			double v = Double.parseDouble(value);

			return new CurrencyAmount(v);
		} catch (NumberFormatException e) {
			throw new CurrencyAmountFormatException(e.toString());
		}
	}

}

File: proj/WinStore_FamilyAccounting/WinStore_FamilyAccountingCore/Utilities/CurrencyAmount.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using WinStore_FamilyAccountingCore.Exceptions.FormatExceptions;

namespace WinStore_FamilyAccountingCore.Utilities
{
    /// <summary>
    /// currency amount ~.2
    /// </summary>
    public class CurrencyAmount : IComparable<CurrencyAmount>
    {
        private int _value;

        /// <summary>
        /// Constructor with ZERO
        /// </summary>
        public CurrencyAmount()
        {
            _value = 0;
        }

        /// <summary>
        /// Construct with input value
        /// </summary>
        /// <param name="value"></param>
        public CurrencyAmount(Double value)
        {
            Set(value);
        }

        /// <summary>
        /// Construct with other amount
        /// </summary>
        /// <param name="amount"></param>
        public CurrencyAmount(CurrencyAmount amount)
        {
            if (amount == null)
            {
                _value = 0;
                return;
            }

            _value = amount._value;
        }

        /// <summary>
        /// this += amount
        /// </summary>
        /// <param name="amount"></param>
        public void AddTo(CurrencyAmount amount)
        {
            _value += amount._value;
        }

        /// <summary>
        /// this -= amount
        /// </summary>
        /// <param name="amount"></param>
        public void MinusTo(CurrencyAmount amount)
        {
            _value -= amount._value;
        }

        /// <summary>
        /// set amount
        /// </summary>
        /// <param name="amount"></param>
        public void Set(CurrencyAmount amount)
        {
            if (amount == null)
            {
                return;
            }

            _value = amount._value;
        }

        /// <summary>
        /// set amount
        /// </summary>
        /// <param name="value"></param>
        public void Set(double value)
        {
            double v = value * 100;
            if (v >= 0)
            {
                _value = (int)(v + 0.5);
            }
            else
            {
                _value = (int)(v - 0.5);
            }
        }

        /// <summary>
        /// Compare to
        /// </summary>
        /// <param name="arg0"></param>
        /// <returns></returns>
        public int CompareTo(CurrencyAmount arg0)
        {
            return this._value - arg0._value;
        }

        /// <summary>
        /// is zero
        /// </summary>
        /// <returns></returns>
        public bool IsZero()
        {
            return this._value == 0;
        }

        /// <summary>
        /// is negative
        /// </summary>
        /// <returns></returns>
        public bool IsNegative()
        {
            return this._value < 0;
        }

        /// <summary>
        /// this = -this
        /// </summary>
        public void Negate()
        {
            this._value = 0 - this._value;
        }

        /// <summary>
        /// ret = oper1 + oper2
        /// </summary>
        /// <param name="oper1"></param>
        /// <param name="oper2"></param>
        /// <returns></returns>
        public static CurrencyAmount Add(CurrencyAmount oper1, CurrencyAmount oper2)
        {
            CurrencyAmount cur = new CurrencyAmount();
            cur.AddTo(oper1);
            cur.AddTo(oper2);

            return cur;
        }

        /// <summary>
        /// ret = oper1 - oper2
        /// </summary>
        /// <param name="oper1"></param>
        /// <param name="oper2"></param>
        /// <returns></returns>
        public static CurrencyAmount Minus(CurrencyAmount oper1,
                CurrencyAmount oper2)
        {
            CurrencyAmount cur = new CurrencyAmount();
            cur.AddTo(oper1);
            cur.MinusTo(oper2);

            return cur;
        }

        /// <summary>
        /// Equals
        /// </summary>
        /// <param name="obj"></param>
        /// <returns></returns>
        public override bool Equals(Object obj)
        {
            CurrencyAmount amount = obj as CurrencyAmount;
            if (amount == null)
            {
                return false;
            }
            return this._value == amount._value;
        }

        /// <summary>
        /// get hash code
        /// </summary>
        /// <returns></returns>
        public override int GetHashCode()
        {
            return this._value;
        }

        /// <summary>
        /// To string
        /// </summary>
        /// <returns></returns>
        public override String ToString()
        {
            double value = this.ToNumber();
            return String.Format("{0:0.00}", value);
        }

        /// <summary>
        /// to number
        /// </summary>
        /// <returns></returns>
        public double ToNumber()
        {
            return _value / 100.0;
        }

        /// <summary>
        /// parse string to currency amount
        /// </summary>
        /// <param name="value"></param>
        /// <returns></returns>
        /// <exception cref="CurrencyAmountFormatException"></exception>
        public static CurrencyAmount Parse(String value)
        {
            double v;
            bool flag = Double.TryParse(value, out v);
            if (flag == false)
            {
                throw new CurrencyAmountFormatException(value);
            }

            return new CurrencyAmount(v);
        }
    }
}
